Use SettingsManager for boot time settings to enable reading from environment variables

ufundo opened the pull request:
Overview
----------------------------------------
Another go at loading some settings from env variables - this time using SettingsManager / SettingsMetadata in a locked down early boot mode.

The approach I've taken here is:

A) add three new optional keys to settings metadata:
- `load_from_env`  = whether to check env var for this setting value
- `is_constant` = whether this setting corresponds to a php define
- `global_name` the name for the constant / env var (probably something prefixed with CIVICRM_....) 

B) use this meta to read from env vars as well as `$civicrm_setting` global when getting mandatory values
- this means env vars act as overrides if there is any database value for the given setting
- env vars are given precedence over $civicrm_setting values

C) build out the early-boot mode of SettingsManager / SettingsMetadata so that we can load some meta in the early bootstrap (before DB / Extension system are available)
- currently there is a very minimal early-boot mode before you call `useDefaults`, which has no meta and just provides a very small number of defaults using `SystemManager::getSystemDefaults` function 
- now we read meta from a subset of the core `settings` folder files, indicated by using `*.boot.settings.php` filename pattern 
- no hooks are called and only local static cache is used
- this means boot-time setting defaults can be loaded from the meta (rather than the exceptional function)
- we also can check our `global_name` / `load_from_env` / `is_constant` keys to act on them early enough
- Q: is there an appreciable performance cost to reading 3 extra metadata files?

D) add `SettingsManager::bootSettings` function to wrap loading `civicrm.settings.php` file
- we need to do be able to hook in before and after the civicrm.settings.php file to define php constants from env vars before any define statements in the settings file itself, and then to define any defaults after
- this function is only used for Standalone in the PR so far (see change to the civicrm.standalone.php template) - I _think_ the equivalent should be reasonably straightforward for other UFs?

E) add settings meta for `CIVICRM_DB_[HOST,USER,...]`
- demonstrate the new functionality with new constant settings meta for the civicrm database connection
- I also added interpolation of DSN from components as part of the calculated values hook in SettingsBag - I think with env vars people will really want to be able to set the components individually without having to calculate the DSN string each time

Before
--------------------
- no loading from env vars
- no metadata for constants
- defaults for system settings are specified separately from all their other meta

After
-------------------
- metadata for constants is available through `Civi::settings()`
- database credentials can be provided to standalone using combo of `CIVICRM_DB_HOST` `CIVICRM_DB_PORT`  `CIVICRM_DB_NAME` `CIVICRM_DB_USER` `CIVICRM_DB_PASSWORD` env vars.
- `CIVICRM_SETTING_EXTENSIONS_DIR` and `CIVICRM_SETTING_EXTENSIONS_URL` can be set using environment variables

ufundo committed:
SettingsMetadata - allow limited use during system bootstrap

ufundo committed:
SettingsManager - merge in env var values to mandatory values

ufundo committed:
SettingsManager - add function to wrap loading settings file

ufundo committed:
SettingsManager - set constants from env settings at boot

ufundo committed:
add civicrm database connection settings meta

ufundo committed:
SettingsBag - interpolate between DSN and Host/Name/User/Pass settings

ufundo committed:
SettingsMetadata - implement local caching for boot settings metadata

ufundo committed:
SettingsManager - set all constants during bootSettings

ufundo commented:
Oof thats a lot of fails.

ufundo committed:
SettingsManager - flatten useDefaults and bootOnly; update comments

ufundo committed:
SettingsBag - throw Exception if trying to set a constant value in the db layer

ufundo commented on settings/Database.boot.setting.php:
```diff
@@ -0,0 +1,124 @@
+    'help_text' => NULL,
+    'is_constant' => TRUE,
+    'load_from_env' => TRUE,
+    'fqn' => 'CIVICRM_DB_NAME',
```
'fully_qualified_name' ? something else?

(used for php constant names and env variable names)

colemanw replied:
Hopefully we can avoid bikeshedding it too much, but we tend to stay away from acronyms and abbreviations due to some bad experiences in the past (ahem, `civicrm_uf_...`). Maybe just `full_name` if that's clear?

ufundo replied:
I was reaching for something with a bit more indication of how/why different from 'name'... but I also dislike abbreviations. `global_name` ?

colemanw replied:
`global_name` sounds fine

ufundo committed:
SettingsManager - swallow errors loading settings file (like
civicrm-drupal)

ufundo commented:
Oh sorry, I forgot about the style fails!  :facepalm: 

(Linter is broken for me atm, for reasons unknown...)

ufundo commented on Civi/Core/SettingsManager.php:
```diff
@@ -328,38 +394,60 @@ public function flush() {
+    foreach ($bootConstants as $key => $meta) {
+      $fqn = $meta['global_name'] ?? NULL;
+      $value = $envSettingsValues[$key] ?? NULL;
+      if ($fqn && !defined($fqn) && !is_null($value)) {
+        define($fqn, $value);
+      }
```
This was just me being crash-averse... but yes something is probably not right if the constant is already defined - so maybe we should crash.

Or throw a warning message? Though we're too early to use any regular warnings.

Happy to remove if preferred.

ufundo committed:
avoid repeat dirname

Co-authored-by: Rich Lott <artfulrobot@users.noreply.github.com>

ufundo commented on Civi/Core/SettingsBag.php:
```diff
@@ -462,4 +479,70 @@ public static function getContributionInvoiceSettingKeys(): array {
+      $componentValues[$componentKey] = $value;
+    }
+
+    $dsn = "mysql://{$componentValues['user']}:{$componentValues['password']}@{$componentValues['host']}:{$componentValues['port']}/{$componentValues['name']}?new_link=true";
```
Good catch!

Do you know if it's a straight url-encode or some specific kind of escaping in this dsn-context?

(There was something with dsn encoding on Joomla recently - scanning this makes me think `url_encode` is prob the surest way forward. @totten / @seamuslee001 may know?)

ufundo replied:
Added `urlencode` and tested with a password containing `@` and `#` and seems to work ok. Wondering if there are any other gotchas here...

totten replied:
> (@ufundo) Added urlencode and tested with a password containing @ and # and seems to work ok. Wondering if there are any other gotchas here...

I haven't fully grokked the patchset/dataflow yet, but just low-level comments on this function...

1. I suggest giving a docbock with example outputs. When juggling between `parse_url()`, `DB::parseDSN()`, `urlencode()`, and `urldecode()`, it's very easy to get tripped-up, and it helps to have the declaration of True North. My guess is that the intended output is like:

    ```php
     /**
      * @return array
      *   Ex: [
      *     'civicrm_db_dsn' => 'mysql://my_user%21:pass%23word@host.name/db_name',
      *     'civicrm_db_user' => 'my_user!',
      *     'civicrm_db_password' => 'pass#word',
      *     'civicrm_db_host' => 'host.name',
      *     'civicrm_db_database' => 'db_name',
      *   ]
      */
    ```

    Esp: indicate which return values are supposed to be encoded-vs-decoded.

    Other parts of the patchset declare the setting `civicrm_db_password`, so my guess is that `civicrm_db_password` is supposed to be plaintext (`pass#word` not `pass%23word`).

2. This function appears to have two main code-paths (one starting from a single DSN-string; the other starting from multiple-properties). Make sure that both code-paths produce the same kind of output.

3. If you're looking at `civicrm.settings.php` and the content of `define('CIVICRM_DSN', '...')`, the format is historically muddled. When reading this data, it's probably better to use `DB::parseDSN()` rather than `parse_url()`.

   * `parse_url()` *only* reads RFC-compliant URLs.
    * `DB::parseDSN()` reads RFC-compliant URLs *as well as* legacy (non-compliant) URLs.
    * If `civicrm.settings.php` originates in the past, then `define('CIVICRM_DSN', '....')` may store an encoding that is not RFC-compliant. (A password like `ab#cd` may be written as `'mysql://username:ab#cd@hostname/dbname'` -- where the literal `#` is non-compliant/ambiguous.)
    * If `civicrm.settings.php` originates recently, then `define('CIVICRM_DSN', '....')` stores an RFC-compliant URL.
    * "Past" vs" "recent" ==> I'm not certain exactly where the dividing line is for Civi-Drupal or Civi-WP, but we had a gnarly+destructive issue circa 5.66-5.73 when Civi-Joomla started to use `parse_url(CIVICRM_DSN)`. (https://lab.civicrm.org/dev/core/-/issues/5128) This issue was fixed by using `DB::parseDSN()` instead.
 
4. Additionally, note that `parse_url()` and `DB::parseDSN()` have slightly different return values:

    * `parse_url()` outputs `['pass' => 'pass%23word']`,
    * `DB::parseDSN()` outputs `['password' => 'pass#word']`

    My guess is that that `parseDSN()` is also closer here -- and `parse_url()` needs extra calls to `urldecode()`. (*But that depends on the declared/intended return-format...*)

ufundo replied:
Have switched to `parseDSN` and updated the docbloc now

ufundo commented:
> is there performance cost to the filesystem scan and glob though

Interesting! I did wonder if all the boot settings should go in `Boot.setting.php` as opposed to a number of `*.boot.setting.php` . It feels like there's quite a lot for one file but it would save the globbing.

ufundo commented:
> DSN: we have to provide all the separate bits in order for it to be accepted, so there's no room for defaults (e.g. "localhost" for host and 3306 for port). I'm fine with this, but it feels like it might catch someone out.

I thought https://github.com/civicrm/civicrm-core/blob/ff173db23a1c4a6fe3567a73418d837d86333227/Civi/Core/SettingsBag.php#L530 would pick up the defaults for components other than password (especially 3306 for the port) - and I thought that was working in my testing... but maybe I've broken it...

ufundo commented:
> Re more testing, I've not done much yet. I may have misunderstood, but I tried running a civibuild with CIVICRM_SETTING_EXTENSIONS_DIR set, and it does not appear to have been consulted at all. 

So the Standalone installer sets a bunch of its own overrides for folder structure currently. But post-install it should pick up your env var and start using it?

It's a good point though - I can look at how this approach would integrate with the installer.

ufundo commented on Civi/Core/SettingsManager.php:
```diff
@@ -239,13 +280,52 @@ protected function getDefaults($entity) {
+    foreach ($specs as $key => $spec) {
+      $fqn = $spec['global_name'] ?? NULL;
+      if ($fqn) {
+        $envValue = getenv($fqn);
+        if ($envValue) {
+          $settings[$key] = $envValue;
+        }
```
Yes.

At this point I think that shouldn't cause trouble, because all the `load_from_env` settings should be truthy strings. But in building out it would be good to add some more intelligent casting / handling. But there's decisions to be made there, which I thought would be easiest to leave out for now. (In the previous PR I had (set FALSE for "0" "false" "FALSE" //  ignore "" // set the string value otherwise. It might be useful that we can consult the "type" from the settings meta at this point.)

totten replied:
* +1 for allowing literal `"0"` or `0` to be respected as concrete values (*booleans*).
    * Since `getenv` [returns `FALSE` for undefined values](https://www.php.net/getenv), this just needs to check  `$envValues !== FALSE` 
    * This makes it *possible* to specify booleans as true or false (`SOME_VAR=1` or `SOME_VAR=0`).
    * We don't need to hold the whole PR on this subissue. (It can be a small/quick follow-up.)
* +1 for leaving out handling for `"true"` / `"false"` -- at least for now. There's sufficient expressiveness with `"0"`/`"1"`, and this allows a *simple* value-mapping.
    * Of course, it's not perfect. Some percentage of sysadmins would feel more comfortable with `SOME_VAR=true`/`SOME_VAR=false` than with `SOME_VAR=1`/`SOME_VAR=0`.
    * If we really want to sugar-coat the env-vars, then I think that moves us to type-dependent parsing... `MY_BOOL=false` and `MY_STR=false` need to parse differently.
    * I don't have a full opinion on whether it's worth the complication, but...
    * IMHO it's safe to add that later. We're not painted into a corner. (*Any env-vars that are legit under today's regime should be equally legit under a future/type-dependent regime.*)

ufundo committed:
SettingsBag - encode dsn components and improve comment

totten commented:
> > > Q: is there an appreciable performance cost to reading 3 extra metadata files?
> >
> > Probably not. I expect the OS's file cache would keep them hot. But is there performance cost to the filesystem scan and glob though?
> > ...I feel like that's probably more of an overhead?...
>
> Interesting! I did wonder if all the boot settings should go in Boot.setting.php as opposed to a number of *.boot.setting.php . It feels like there's quite a lot for one file but it would save the globbing.

I would wager there is no appreciable cost or difference for an extra metadata file or an extra glob().

* In typical usage on a VPS/dedicated web-host, both should be covered by cache layers (PHP opcode cache and OS/[V]FS cache, respectively). Both should involve relatively small# operations.
* Things change if multiplied hundreds/thousands of times.
    * Searching an `include_path` with 100 folders (on every pageload) would be alarming. (*IMHO, this is where file-scans have gotten a bad-rep.*)
    * Reading 100 metadata files (on every pageload) would be alarming.
* Things also change if the underlying caches perform badly (e.g.  an anemic amount of cache-space or an expensive network file-system).  But if a deployment does have poor caching, then they'd probably feel it in other ways (regardless of this change).  In any case, conventional-wisdom pushes deployers to models where this is unlikely to be a problem (eg VPS/dedicated).

IMHO, either approach (read one `Boot.setting.php` or read 5x `*.setting.php` files) should be within bounds. (*One is slightly faster; the other slightly more organized; both are sufficiently fast and organized.*)

totten commented on Civi/Core/SettingsManager.php:
```diff
@@ -328,38 +394,60 @@ public function flush() {
+      $error = include_once $settingsPath;
+      if ($error === FALSE) {
+        echo "Could not load the settings file at: {$settingsFile}\n";
+        exit();
```
1. `$settingsPath` vs `$settingsFile`? Copy-paste error?
2. Looking at the places which load `civicrm.settings.php`, there appear to be mixed signals about whether/how to check this output. (*Lightly-edited stream-of-conscious...*)
    *  Survey
        * __Regular/Foreground__: In the regular routing (for Drupal/Backdrop/WordPress), the init function simply says `return FALSE`. 
            * (*Recoverable error; but no obvious trace/exit-code*)
        * __Regular/Foreground__: In the regular routing (for Joomla), it doesn't have any check like this. 
            * (*It simply proceeds; no error*)
        * __Regular/Foreground__: In the regular routing (for Standalone), it didn't have a check like this. (But this PR adds the `echo`+`exit`.)
        * __Extern/Background__: In the `civicrm.config.php` templates (for Drupal/Backdrop/Joomla), it's like the above (`echo`+`exit`).
            * (*Non-recoverable error; no backtraces; misleading exit-code*)
        * __Extern/Background__: In the `civicrm.config.php` template for WordPress, and in `cv`'s original `Bootstrap.php`, it throws an `Exception`.
           * (*Recoverable error; backtraces and exit-codes are better*)
    * Debugging this kind of `exit()` can be tricky, so I favor only as a last-resort. Survey shows... you don't have to have the `exit()`.
    * This guard likely originates circa PHP4 -- before `try`/`catch`/`throw`/`Exception` existed. 
    * My guess is that `civicrm.config.php.XXXX` did this because they had un-guarded calls to `include_once` -- and this pattern provided a way to give a prettier error-message.  But in this context, you have the `file_exists()` guard (L428). So I don't think you'll ever get a meaningful error-signal on this.
        * (*In theory, you might say: "Well, what if `civicrm.settings.php` has a syntax error or a bad method-call?" That could happen... but this pattern here won't catch that error...*)

__tldr__: Aaaah! `exit()`!!! Aah!

ufundo replied:
I took this from the Drupal `civicrm.config.php` because I thought that was somewhat canonical, but will remove if it is intended for PHP4!

ufundo replied:
(FWIW my assumed rationale was "dont leak anything to web if civicrm.settings.php has a syntax error or a bad method-call"?)

ufundo replied:
Have removed.

ufundo committed:
SettingsBag interpolateDsnSettings - use DB::parseDSN and add examples

ufundo committed:
stop trying to capture any settings file error

ufundo commented:
>IMHO, either approach (read one Boot.setting.php or read 5x *.setting.php files) should be within bounds. (One is slightly faster; the other slightly more organized; both are sufficiently fast and organized.)

My preference is for organised (i.e. what is here)

totten commented on tests/phpunit/Civi/Core/SettingsManagerTest.php:
```diff
@@ -50,7 +50,7 @@ public function testTwoDomains(): void {
     $da = \CRM_Core_DAO::createTestObject('CRM_Core_DAO_Domain');
     $db = \CRM_Core_DAO::createTestObject('CRM_Core_DAO_Domain');
```
(*This comment has nothing to do with this line of code. But I want to put my comments as resolvable subthreads, and it has to do with testing, so... it goes on a test file.* 🙃)

The upgrade test failure seems to be pretty consistent.  The errors aren't immediately visible in the test-console, but you can find see them in the "Test report", eg [here](https://test.civicrm.org/job/CiviCRM-Test/18445/testReport/). (FWIW, I get [similar results](https://gist.github.com/totten/66ed9f1e0e37e205bab6b93880f62908).)

To reproduce it, one can execute:

```
civibuild ut dmaster '5.13.3*'
```

(That command basically just loads https://github.com/civicrm/civicrm-upgrade-test/blob/master/databases/5.13.3-multilingual_af_bg_en.mysql.bz2 into CiviCRM's MySQL database and runs `cv updb`.)

ufundo replied:
Yep I clicked "retry" on that in vain. It's weird to me... but definitely a thread worth pulling. I'll have a look. Thanks for the pointers :+1:

totten commented on Civi/Core/SettingsManager.php:
```diff
@@ -328,38 +394,56 @@ public function flush() {
+      if (!defined('CIVICRM_SETTINGS_PATH')) {
+        define('CIVICRM_SETTINGS_PATH', $settingsPath);
+        require_once $settingsPath;
+      }
```
@ufundo  When I do `civibuild reinstall` on my `standalone-clean` site, I get a crash. The main install step runs:

```bash
cv core:install ...
```

But then calls to `cv` fail:

```bash
cv ev 'return "hello world";'

cv api3 system.get

cv ext:list -L

cv ev 'echo CIVICRM_SETTINGS_PATH;'
## ^^ This is actually the next step of `civibuild reinstall`, but it's coincidental.
```

Specifically, the commands fail at:

```
[PHP Warning] require_once(/CRM/Core/ClassLoader.php): failed to open stream: No such file or directory at phar:///Users/totten/bknix/bin/cv/lib/src/Bootstrap.php:93

Fatal error: require_once(): Failed opening required '/CRM/Core/ClassLoader.php' (include_path='.:/Users/totten/bknix/build/s3/web/core:/Users/totten/bknix/build/s3/web/core/packages:/Users/totten/bknix/build/s3/web/core/vendor/tecnickcom:/Users/totten/bknix/build/s3/web/core/vendor/pear/auth_sasl:/Users/totten/bknix/build/s3/web/core/vendor/pear/console_getopt:/Users/totten/bknix/build/s3/web/core/vendor/pear/db:/Users/totten/bknix/build/s3/web/core/vendor/pear/log:/Users/totten/bknix/build/s3/web/core/vendor/pear/mail:/Users/totten/bknix/build/s3/web/core/vendor/pear/mail_mime:/Users/totten/bknix/build/s3/web/core/vendor/pear/net_smtp:/Users/totten/bknix/build/s3/web/core/vendor/pear/net_socket:/Users/totten/bknix/build/s3/web/core/vendor/pear/pear-core-minimal/src:/Users/totten/bknix/build/s3/web/core/vendor/pear/pear_exception:/Users/totten/bknix/build/s3/web/core/vendor/pear/validate_finance_creditcard:.:/nix/store/sx8q6l82wbdzhkqbyapka4f5pwncijwl-php-7.4.29/lib/php') in phar:///Users/totten/bknix/bin/cv/lib/src/Bootstrap.php on line 93
```

The problem?

* `$civicrm_root` is undefined.... because `civicrm.settings.php` hasn't loaded at all. 
* Recall https://github.com/civicrm/cv/pull/193/commits/3837693b1d7e27eb95d0e598461e3523c8b4c028 -- it changed `findCivicrmSettingsPhp()` to treat `civicrm.standalone.php` as a work-a-like for `civicrm.settings.php`.
*  As a consequence, `Bootstrap.php` puts `/path/to/my/civicrm.standalone.php` [into `CIVICRM_SETTINGS_PATH`](https://github.com/civicrm/cv/blob/v0.3.51/lib/src/Bootstrap.php#L211).
* Now this code (`SettingsManager::bootSettings()`) considers loading the real `civicrm.settings.php`... and declines.
* So we never load the real `civicrm.settings.php`.

Why did it work before? Because it didn't matter if `CIVICRM_SETTINGS_PATH` was slightly munged --  [it proceeded to load anyway](https://github.com/civicrm/civicrm-core/blob/5.76/setup/res/civicrm.standalone.php.txt#L28).

My first resolution was https://gist.github.com/totten/5d71c4b4a01e2f1678b3929265c5ab67, but that feels like putting a hack on top of a hack. I'm not sure pretty way to do that.

Perhaps this would be cleaner: make `SettingsManager::bootSettings()` behave more like the prior revision of `civicrm.standalone.php.txt`? i.e.

```diff
diff --git a/Civi/Core/SettingsManager.php b/Civi/Core/SettingsManager.php
index 54b1146a27..5e92566e16 100644
--- a/Civi/Core/SettingsManager.php
+++ b/Civi/Core/SettingsManager.php
@@ -428,8 +428,8 @@ class SettingsManager {
     if (file_exists($settingsPath)) {
       if (!defined('CIVICRM_SETTINGS_PATH')) {
         define('CIVICRM_SETTINGS_PATH', $settingsPath);
-        require_once $settingsPath;
       }
+      require_once $settingsPath;
     }

     // get all effective values from the settings bag (resolving defaults etc)
```

totten commented on settings/Database.boot.setting.php:
```diff
@@ -0,0 +1,124 @@
+    'help_text' => NULL,
+    'is_constant' => TRUE,
+    'load_from_env' => TRUE,
+    'global_name' => 'CIVICRM_DSN',
```
> (@ufundo ) edit civicrm.settings.php - instead of using the PHP constant set environment variables... And site still works...
>
> Potential gotcha at this stage: if you have `if (!defined('CIVICRM_DSN')) define('CIVICRM_DSN', );` in your civicrm.settings.php, this will jump in and set the DSN explicitly before the interpolation from components happens.

OK, cool. So riffing on this suggested experiment, I edited `CIVICRM_DSN` to specifically to put in *bad DSN* -- and then called Civi with an env-var containinga good DSN:

```bash
CIVICRM_DSN="mysql://s3civi_jq9sa:XXXX@127.0.0.1:3307/s3civi_jprre?new_link=true" \
  cv ev 'echo 123;'
```

And it worked with the env-var. 😄 

Similarly, I tried piece-wise credentials. (And per your  hint, I removed the `define()`.) 

```bash
CIVICRM_DB_USER=s3civi_jq9sa \
  CIVICRM_DB_PASSWORD=XXX \
  CIVICRM_DB_HOST=127.0.0.1 \
  CIVICRM_DB_PORT=3307 \
  CIVICRM_DB_NAME=s3civi_jprre  \
  cv ev 'echo 123;'
```

This also worked. 🎉 

(*All the above required the patch for `SettingsManager::bootSettings()` to work with cv.*)

ufundo replied:
Hurrah :grinning:

totten commented on Civi/Core/SettingsMetadata.php:
```diff
@@ -49,8 +59,30 @@ class SettingsMetadata {
    *   - pseudoconstant
+   *   - global_name (name for php constant / environment variable corresponding to this setting)
+   *   - load_from_env (whether this setting should be read from corresponding environment variable)
+   *   - is_constant (whether a PHP constant should be defined for this setting)
```
Nitpick: The naming feels asymmetric. Any of these would feel more consistent:

* (A) For one string, two bools
    * (A1) `string global_name` + `bool is_constant` + `bool is_env_var`
    * (A2) `string global_name` + `bool use_constant` + `bool use_env_var`
    * (A3) `string global_name` + `bool load_constant` + `bool load_env_var`
    * (Between these, `is_*` has some precedent within setting-metadata, and it's closer to current names. `use_*` and `load_*` seem OK if they at least match each other, but fwiw I don't see as much precedent.)
* (B) For two strings
    *  (B1)`string constant' + `string env_var`
* (C) For one string, one array
    * (C1) `string global_name` + `array sources`
    * (List of supported sources -- php-constants, environment, database, php-settings, or whatever)

The smallest revision would be to go to (A1). And it would be more consistent. 😄 

(*(B) still looks cleaner imho. But it's subjective, and I don't want to block the whole changeset on that. Still, for a couple months, these will just be internal keys... so if you see the light later, we could still change it...*  😉)

(*(C) has its merits, but it's slippery to implement.*)

ufundo replied:
So found this interesting because I don't like B much at all...

For me the `is_constant`ness is quite different from the `whether the value can load_from_env var`ness. 

The latter I thing is very tightly saying "the value for this setting *might* come from the environment, so check for it". IMHO it should definitely only be about the source of the value. `putenv()` doesn't belong anywhere in the codebase.

`is_constant` is a bit more slippery but I think really it should mean something like is "the value of this setting must not change during a given request".  It so happens that currently we use php-constants to implement this, because they behave in this way. And given there are all sorts of pre-existing places where php-constants directly  written to / read, if we're in the business of managing settings, we have to use them as both source and sink (i.e. checking for pre-existing values, and then making sure we define them for any values we know). 

That's maybe a bit abstract - in practical terms I think there may be settings which are reasonable to `load_from_env` but needn't `be_constant`. Slightly random example but take `menubar_color` -- useful to be able to override with an env var on staging, but I don't think it needs to be a constant?

ufundo replied:
Sorry, the other piece to that: I think part of what I don't like with `is_env_var` is that it suggests it is *always* an env var - but I'm imagining plenty of things that might be loadable from the environment without having to be.

Would `is_env_loadable` make everyone happy?

ufundo replied:
Whilst we're on naming, I was musing on my cycle to work today whether `fully_qualified_name` would actually be better than `global_name` .

I'm worried about potential misidentification with `$_GLOBAL` variables, which is a whole nother thing...

totten committed:
SettingsManager::bootSettings() - Match prior behavior. Fix cv.

totten committed:
SettingsManager::getEnvSettingValues() - Treat 0 as a legal value

Here's how I tried the change:

* In `settings/`, reconfigure `debug_enabled` to enable env-var.
* Ensure that `debug_enabled` is not assigned via civicrm.settings.d or MySQL
* Alternately run these commands:

```
CIVICRM_DEBUG=0 cv vget /debug/
CIVICRM_DEBUG=1 cv vget /debug/
```

totten committed:
SettingsMetadata - Rename new property load_from_env to is_env_var

This makes it more consistent with other properties in the same scope.

totten commented:
I think it would be great if we could get this merged at the start of the upcoming `5.78.alpha1` cycle.

Summary of suggestions:

1. Treat `MY_VAR=0` as a legal value ([discussion](https://github.com/civicrm/civicrm-core/pull/30533#discussion_r1662751168))
2. Fix compatibility with `cv` (suggested [patch](https://github.com/civicrm/civicrm-core/pull/30533#discussion_r1706719942))
3.  Rename `load_from_env` to `is_env_var`  (or anything [similar](https://github.com/civicrm/civicrm-core/pull/30533#pullrequestreview-2223983012) to make more consistent names)
4. Get the remaining tests to pass (eg [upgrade test](https://github.com/civicrm/civicrm-core/pull/30533#pullrequestreview-2223394232))
5. Finally, add the `run-extended-tests` label and check results (since this touches some low-level functionality)

I'll push up 1,2,3 because they're pretty small - and that will get us test-results a bit sooner. But I'll the subdiscussions for @ufundo to review. With luck, that will just leave 4+5.

ufundo commented:
Hey, thanks @totten :)

Sorry saw comments but didnt have time to respond today... I agree completely with 1 and 2 but I actually disagree on 3. I think those things are rightly asymmetric. Will try explain tomorrow!

ufundo committed:
SettingsMetadata - rename is_env_var to is_env_loadable

michaelmcandrew commented:
Hey all,

Sorry to be late to the party on this one.

A couple of questions from me, around intended usage and the installation process. I'm thinking of this in the context of a typical Docker use case.

At the core of many/most web application docker workflows is something like the following:

1. Download an application image, let's say `civicrm/standalone:5.76.0`.
2. Run the image with some environment variables supplied (e.g. database credentials, etc.)
3. Check whether the system has been installed and boot as normal if so.
4. If the system has not been installed, either:
  a. Run an installation process.
  b. Display a 'This site has not been installed' message.

So, some  questions based on the above:

1. How does CiviCRM currently detect whether a site has been installed? Is it based on the presence of a settings file? Or is it attempting to connect to a database and executing a query? Or something else? Is there a `cv status` command that we could call or similar before starting apache in the entrypoint script or similar.

2. Now that we can pass in database connection parameters (and other settings) via environment variables, do we still need to have a civicrm settings file? (and if so, is there a default one that we can bake into the image?) Hopefully that is the case. Then all we will have to do is pass the DB credentials to the container via environment variables.
 
3. The current CLI command that we have for installing is `cv core:install`. I think I am right in assuming that creating a settings file is baked into this command. It probably makes sense to have an option for the installer to skip attempting to create a settings file and just create the database tables and any other necessary files.

Also, while I am here, it probably makes sense to talk about upgrades. Many images also check for any outstanding database migration scripts as part of their entrypoint (which is run when the container starts). We can potentially run `cv upgrade:db` as part of the entrypoint script as well. We might want to make sense to make this configurable.

michaelmcandrew commented:
Also @ufundo - do you have any suggestions about how I could test this new image as part of the docker image we are creating? Sorry - can't quite get my head around how I would do that.

ufundo commented:
> do you have any suggestions about how I could test this new image as part of the docker image we are creating? 

It's tricky as the PR branch doesn't get packaged into a release automatically. But I think you could have something like this in a Dockerfile:

```
RUN curl -OL https://github.com/ufundo/civicrm-core/archive/refs/heads/boot-settings.zip && unzip boot-settings.zip && mv civicrm-core-boot-settings core
RUN composer install -d core
RUN php core/tools/standalone/bin/scaffold .
RUN [apply patch to civicrm.standalone.php]
``

If merged, all the above could be replaced with
```
RUN curl -OL https://download.civicrm.org/civicrm-5.76.1-standalone.tar.gz  | tar xzvf
RUN [apply patch to civicrm.standalone.php]
```

michaelmcandrew commented:
Thanks @ufundo for both of these responses. I think I have what I need to make a bit more progress on the image now and should be able to test out your branch. Will see how it goes and report back.

michaelmcandrew commented:
@ufundo - I have switched to using the tarball installation method for now.

I have also added a method for creating an image using a nightly tarball from this page: https://download.civicrm.org/latest/. See https://lab.civicrm.org/michaelmcandrew/civioneclick-docker/-/tree/aba2d142796bd69b1af5a533dd81fe46631870d6#development-and-contributing.

Hopefully, once https://github.com/civicrm/civicrm-core/pull/30835 is merged, I can find a way apply a 'patch' on the tarball release and from this PR (similar to what you suggested above) and test this.

ufundo commented:
For the upgrade test, the offending query seems to be
```
"SELECT  v.label as label ,v.name as value, v.grouping as `grouping`FROM   civicrm_option_value v,       civicrm_option_group gWHERE  v.option_group_id = g.id  AND  g.name            = 'languages'  AND  g.is_active       = 1  AND  v.is_active = 1  AND ( v.component_id IS NULL  OR v.component_id IN (SELECT id FROM civicrm_component WHERE name IN ("CiviEvent","CiviContribute","CiviMember","CiviMail","CiviReport","CiviPledge")) )  ORDER BY v.weight"
```
which is triggered when getting the language options, when flushing the cache, after enabling Flexmailer at the end of the install. 

Weird.

Just dropping here for now, will look more on Friday.

ufundo commented:
Specifically it says ```"1054 ** Unknown column 'v.label' in 'field list'"```

My first thought was is the DSN being dropped at some point during the settings bootstap --- but it looks fine in the `DB_common` object.

ufundo commented on Civi/Core/SettingsManager.php:
```diff
@@ -153,7 +194,7 @@ public function getBagByDomain($domainId) {
     if (!isset($this->bagsByDomain[$domainId])) {
       $this->bagsByDomain[$domainId] = new SettingsBag($domainId, NULL);
-      if (\CRM_Core_Config::singleton()->dsn) {
+      if (!$this->preBoot) {
```
^ change that flattens the "third" state into the two explicit states

ufundo committed:
SettingsManager - fully boot before applying locale in CRM_Core_Config

ufundo commented:
Ok. Definitely a string worth pulling!

The issue seems to be that this line https://github.com/civicrm/civicrm-core/blob/5889153e3995d99da9029070bd6cd8158662ec07/CRM/Core/Config.php#L103 is relying on the subtle third state of `SettingsManager`. `useDefaults` has not been called yet, so it's not the full settings manager. But `dsn` has been set, so it can and does read database setting values (in particular `lcMessages` and `languageLimit`). 

In the PR, I was trying to flatten this third state into the two explicit ones ( preBoot, postBoot ) with [this change](https://github.com/civicrm/civicrm-core/pull/30533/files#r1718188602)

But that's borking the applyLocale call, because it comes before `bootComplete()` is called in `CRM_Core_Config::singleton`.

**Option A: revert the change to the gate on `loadValues`** - probably the quickest fix, but retains this third "in between" state, which I don't much like the feel of

**Option B: move `bootComplete` up a few lines** - this feels preferable to me to keep clarity of SettingsManager state -- but may have knock ons. It would move above the `hook_civicrm_config` - I don't know how pertinent that is? My read of the existing SettingsManager code is our primary concerns before we call `useDefaults` are that we have the DB and extensions are booted, so they can modify the settings metadata. I think that has happened once the container is loaded above.

**Option C: move `applyLocale` down a few lines** - this feels like fixing the symptom rather than the cause...

Testing locally B + C both fix the upgrade test.

I'm going to try pushing B to see if what other tests say.

ufundo commented:
Ok, so that caused these: https://test.civicrm.org/job/CiviCRM-Test-QLow/56480/testReport/

Which seems to be because the `en_US` default for `lcMessages` is loaded, and something then decides the database is "localised" to `en_US`, and it tries to look in `option_value_en_US` tables. Which is kinda logical I guess.

It seems like a few places in early boot are relying on the fact that `\Civi::settings()->get('lcMessages')` doesn't yet return the en_US . I.e. from their perspective the default is NULL.

Making the default NULL is probably going to break other things, but let's see. 

Maybe for this exceptional case, we could add `if is_null(\Civi::setttings()->get('lcMessages')) \Civi::settings()->set('lcMessages', 'en_US');` a little bit later in the bootstrap?

ufundo committed:
applyLocale - dont apply invalid lcMessages setting on multilang

ufundo commented:
> Making the default NULL is probably going to break other things, but let's see.

And so it proved.

Instead, I've dug into `applyLocale`. The issue was coming from getting a value ('en_US') from `Civi::settings()->get('lcMessages')` that wasn't one of the available locales on multilang.

This seems like a good thing to validate anyway -- in the current bootstrap it will check the database value at this point, and if the database value isn't one of the available locales, it will cause a nasty crash.

So I've added a gate to clear it. Feeling quite hopeful for this test run :crossed_fingers:

ufundo commented:
Ok, basic tests all passed except for a trailing space :+1: 

Now we try extended.

ufundo commented:
At this point looks to me like most of the fails are also in the `master` baseline for extended tests.

Obviously failing across the board on `standalone` isn't an ideal baseline, so going to try fix that fire and then re-run tests here.

ufundo commented:
Rerunning extended tests now Standalone should be revived on `master`

ufundo commented:
Think all those extended test fails are pre-existing.

@totten I think we've reached step 5. Though think it would be good if you could check the last few commits around multilingual `applyLocale`

michaelmcandrew commented:
@ufundo - I made some improvements to the docker image to test this PR.

- it can now [build nightly images](https://lab.civicrm.org/michaelmcandrew/civioneclick-docker/-/blob/main/build.sh)
- [the docker compose example file uses nightly](https://lab.civicrm.org/michaelmcandrew/civioneclick-docker/-/blob/main/compose.example.yaml#L3) (for now)
- and I have also added [instructions on how to test unmerged PRs](https://lab.civicrm.org/michaelmcandrew/civioneclick-docker/-/tree/main#testing-prs).

Using the following docker compose file, I was able install (using this method https://lab.civicrm.org/michaelmcandrew/civioneclick-docker/-/blob/main/install.sh) and download an extension to the custom extension directory that I had set via an environment variable :tada:

```yaml
services:
  web:
    build:
      dockerfile_inline: |
        FROM civicrm-standalone-nightly
        RUN curl -o pr.diff https://patch-diff.githubusercontent.com/raw/civicrm/civicrm-core/pull/30533.diff; \
          patch -i ../pr.diff -p1 -d core -t; \
          rm pr.diff
    ports:
      - 8760:80
    restart: unless-stopped
    volumes:
      - private:/var/www/html/private
      - ext:/var/civicrm-extensions
    environment:
#      - CIVICRM_DB_HOST=db
#      - CIVICRM_DB_PORT=3306
#      - CIVICRM_DB_NAME=civicrm
#      - CIVICRM_DB_USER=civicrm
#      - CIVICRM_DB_PASSWORD=passwordiswrong
      - CIVICRM_SETTINGS_PATH=/var/www/html/civicrm.standalone.php
      - CIVICRM_SETTING_EXTENSIONS_DIR=/var/civicrm-extensions
  db:
    image: mariadb:10.11
    restart: unless-stopped
    volumes:
      - db:/var/lib/mysql
    environment:
      - MYSQL_ROOT_PASSWORD=insecure
      - MYSQL_DATABASE=civicrm
      - MYSQL_USER=civicrm
      - MYSQL_PASSWORD=password
volumes:
  db:
  private:
  ext:
```

Based on your comments above, I was hoping to be able to uncomment the `#      - CIVICRM_DB_PASSWORD=passwordiswrong` line and get a database connection error but it continued to connect using settings in the settings file.

Deleting the `/var/www/html/private/civicrm.settings.php' file caused CiviCRM to think it was uninstalled.

Deleting the settings file returned me to the installation screen.

So great progress!

But a few kinks to iron out still. Happy to jump on a call to discuss at some point.

ufundo commented:
retest this please

totten commented on CRM/Core/BAO/ConfigSetting.php:
```diff
@@ -147,14 +147,20 @@ public static function applyLocale($settings, $activatedLocales) {
+    }
+
     // When there is a choice of permitted languages.
+    // Q: why would this be different from the locales?
```
(*Tangential subthread about the comment*)

Yeah, that kind of question hurts the brain. [The general problem](https://lab.civicrm.org/dev/translation/-/issues/78) is [this](https://github.com/civicrm/civicrm-core/blob/b37ca98db9a6f055f3f504a77651f9a3d2c36d6c/Civi/Core/Locale.php#L19-L47):

> The software comes with a list of 200 communication-locales (OptionValues), and admins may register more. There are only ~50 locales supported by `ts()` and 1-3 locales in the DB (for a typical business-entity). If you use any of these other locales, then some services must raise errors (or fallback to an alternate locale).

Those numbers are completely different. The situation invites edge-cases. If you operate in these circumstances, then the system has some *de-facto negotiation behavior*. ("We see the user prefers `xx_XX`, but we cannot meet it head-on, so we switch to resource `yy_YY` or throw error `Zzz`.")

Civi has two implementations of negotiation behavior. (*Naturally! Why wouldn't it?*)

* `applyLocale()` identifies the *preferred* locale by examining `$_REQUEST` data and CMS/user data...  which is good.  Then it decides whether/how it can satisfy the preference, but it doesn't try very hard. (It always need a full match in all layers. It doesn't support the `partial_locales` setting.)
* `setLocale()` (aka `Locale::negotiate()`) requires you to tell it the `$preferred` locale. But it [works harder](https://github.com/civicrm/civicrm-core/blob/b37ca98db9a6f055f3f504a77651f9a3d2c36d6c/Civi/Core/Locale.php#L176-L217) to satisfy the user's preference. (It does respect `partial_locales`.)

Perhaps... in an ideal world... we would put these together (eg `applyLocale()` would call `setLocale($preferred)`). But validating that change is another adventure...

ufundo replied:
Thanks @totten . Have dropped a reference to this into the comment.

Anything non-tangential outstanding? :slightly_smiling_face:

ufundo committed:
add comment link to wisdom

ufundo committed:
SettingsManager - export a composed DSN value to define before loading settings file
